enterprise/wsproxy/wsproxysdk/wsproxysdk.go

```diff
@@ -171,18 +202,178 @@ func (c *Client) RegisterWorkspaceProxy(ctx context.Context, req RegisterWorkspa
+// stop immediately and the context error will be returned to the FailureFn.
+//
+// The returned channel will be closed when the loop stops and can be used to
+// ensure the loop is dead before continuing. When a fatal error is encountered,
+// the proxy will be deregistered (with the same ReplicaID and AttemptTimeout)
+// before calling the FailureFn.
+func (c *Client) RegisterWorkspaceProxyLoop(ctx context.Context, opts RegisterWorkspaceProxyLoopOpts) (RegisterWorkspaceProxyResponse, <-chan struct{}, error) {
+	if opts.Interval == 0 {
+		opts.Interval = 30 * time.Second
+	}
+	if opts.MaxFailureCount == 0 {
+		opts.MaxFailureCount = 10
+	}
+	if opts.AttemptTimeout == 0 {
+		opts.AttemptTimeout = 10 * time.Second
+	}
+	if opts.MutateFn == nil {
+		opts.MutateFn = func(_ *RegisterWorkspaceProxyRequest) {}
+	}
+	if opts.CallbackFn == nil {
+		opts.CallbackFn = func(_ context.Context, _ RegisterWorkspaceProxyResponse) error {
+			return nil
+		}
+	}
+
+	failureFn := func(err error) {
+		// We have to use background context here because the original context
+		// may be canceled.
+		deregisterCtx, cancel := context.WithTimeout(context.Background(), opts.AttemptTimeout)
+		defer cancel()
+		deregisterErr := c.DeregisterWorkspaceProxy(deregisterCtx, DeregisterWorkspaceProxyRequest{
+			ReplicaID: opts.Request.ReplicaID,
+		})
+		if deregisterErr != nil {
+			opts.Logger.Error(ctx,
+				"failed to deregister workspace proxy with Coder primary (it will be automatically deregistered shortly)",
+				slog.F("err", deregisterErr),
+			)
+		}
+
+		if opts.FailureFn != nil {
+			opts.FailureFn(err)
+		}
+	}
+
+	originalRes, err := c.RegisterWorkspaceProxy(ctx, opts.Request)
+	if err != nil {
+		return RegisterWorkspaceProxyResponse{}, nil, xerrors.Errorf("register workspace proxy: %w", err)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+
+		var (
+			failedAttempts = 0
+			ticker         = time.NewTicker(opts.Interval)
+		)
+		for {
+			select {
+			case <-ctx.Done():
+				failureFn(ctx.Err())
+				return
+			case <-ticker.C:
+			}
+
+			opts.Logger.Debug(ctx,
+				"re-registering workspace proxy with Coder primary",
+				slog.F("req", opts.Request),
+				slog.F("timeout", opts.AttemptTimeout),
+				slog.F("failed_attempts", failedAttempts),
+			)
+			opts.MutateFn(&opts.Request)
+			registerCtx, cancel := context.WithTimeout(ctx, opts.AttemptTimeout)
+			res, err := c.RegisterWorkspaceProxy(registerCtx, opts.Request)
+			cancel()
+			if err != nil {
+				failedAttempts++
+				opts.Logger.Warn(ctx,
+					"failed to re-register workspace proxy with Coder primary",
+					slog.F("req", opts.Request),
+					slog.F("timeout", opts.AttemptTimeout),
+					slog.F("failed_attempts", failedAttempts),
+					slog.F("err", err),
+				)
+
+				if failedAttempts > opts.MaxFailureCount {
+					failureFn(xerrors.Errorf("exceeded re-registration failure count of %d: last error: %w", opts.MaxFailureCount, err))
+					return
+				}
+			}
```

Comment:
Should there be a continue here?
```suggestion
			if err != nil {
				failedAttempts++
				opts.Logger.Warn(ctx,
					"failed to re-register workspace proxy with Coder primary",
					slog.F("req", opts.Request),
					slog.F("timeout", opts.AttemptTimeout),
					slog.F("failed_attempts", failedAttempts),
					slog.F("err", err),
				)

				if failedAttempts > opts.MaxFailureCount {
					failureFn(xerrors.Errorf("exceeded re-registration failure count of %d: last error: %w", opts.MaxFailureCount, err))
					return
				}
				continue
			}
```